FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -121,12 +127,168 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
+    for (NSString *key in properties) {
+        NSString *suffix = @"";
+
+        // properties should always be a NSDictionary, but still check to make sure the value we get is an object
+        // Check Type Encoding: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html
+        const char* typeCode = @encode(typeof(props[key]));
+
+        if ([@"@" isEqualToString:@(typeCode)]) {
+            NSObject *obj = properties[key];
+            // if it's a NSDictionary: recurrsively get props, and we can not take underscore in key for nested objects, so we simply remove underscores
+            // more info: https://help.fullstory.com/hc/en-us/articles/360020623234-FS-Recording-Client-API-Requirements
+            if ([obj isKindOfClass:[NSDictionary class]]) {
+                [props setValue:[self getSuffixedProps:properties[key]]
+                         forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
+            } else if ([obj isKindOfClass:[NSArray class]]) {
+                NSArray *arr = (NSArray *) obj;
+                NSDictionary *dict = [self getDictionaryFromArrayObject:arr withKey:key];
+                [self appendToDictionary:props withKey:@"" andValue:dict];
+            } else {
+                // if this is not a dictionary or array, then treat it like simple values, if data falls outside of these, we don't try to infer anything and just send as is to the server
+                suffix = [self getSuffixStringFromSimpleObject:obj];
+                [self appendToDictionary:props withKey:[key stringByAppendingString:suffix] andValue:properties[key]];
+            }
+        } else {
+            #ifdef DEBUG
+                NSAssert(FALSE, @"key `%@` is not an object can't be serialized for FS custom event.", key);
+            #else
+                NSLog(@"key `%@` is not an object can't be serialized for FS custom event.", key);
+                // if prod, then ignore this pops all together
+            #endif
+        }
+    }
+
+    return props;
+}
+
+- (NSString *)getSuffixStringFromSimpleObject:(NSObject *)obj {
+    // default to no suffix;
+    NSString * suffix = @"";
+    if ([obj isKindOfClass:[NSNumber class]]) {
+        // defaut to real
+        suffix = @"_real";
+        NSNumber *n = (NSNumber *) obj;
+        const char *type = n.objCType;
+        if ([@"i" isEqualToString:@(type)]) {
+            suffix = @"_int";
+        } else if ([@"B" isEqualToString:@(type)]) {
+            // bool-type gets encoded as number number, and doesn't get encoded into 'B', but check anyway
+            suffix = @"_bool";
+        }
+    } else if ([obj isKindOfClass:[NSDate class]]) {
+        suffix = @"_date";
+    } else if([obj isKindOfClass:[NSString class]]) {
+        suffix = @"_str";
+    }
+
+    return suffix;
+}
+
+- (NSDictionary *)getDictionaryFromArrayObject:(NSArray *)arr withKey:(NSString *)key {
```

Comment:
convert array and flatten them into "primitive" arrays like so:
arr[{key:val1}, {key:val2}] -> key_str:[val1,val2]